File: server/app/schemas/alignment.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


AlignmentMode = Literal["global", "local"]


class AlignmentScoring(BaseModel):
    match_score: int = 1
    mismatch_score: int = -1
    gap_penalty: int = -1


class AlignmentAlignRequest(BaseModel):
    sequence_a: str = Field(..., min_length=1)
    sequence_b: str = Field(..., min_length=1)
    mode: AlignmentMode = "global"
    scoring: AlignmentScoring = Field(default_factory=AlignmentScoring)


class AlignmentResult(BaseModel):
    aligned_a: str
    aligned_b: str
    score: int
    start_a: int
    end_a: int
    start_b: int
    end_b: int
    operations: list[str]


class AlignmentAlignResponse(BaseModel):
    mode: AlignmentMode
    sequence_a: str
    sequence_b: str
    scoring: AlignmentScoring
    result: AlignmentResult

File: client/src/lib/api/alignment.ts
import { useQuery, type UseQueryResult } from "@tanstack/react-query";

import { axiosInstance } from "@/lib/api/client";

export type AlignmentMode = "global" | "local";

export type AlignmentScoring = {
  match_score: number;
  mismatch_score: number;
  gap_penalty: number;
};

export type AlignmentAlignRequest = {
  sequence_a: string;
  sequence_b: string;
  mode: AlignmentMode;
  scoring: AlignmentScoring;
};

export type AlignmentAlignResponse = {
  mode: AlignmentMode;
  sequence_a: string;
  sequence_b: string;
  scoring: AlignmentScoring;
  result: {
    aligned_a: string;
    aligned_b: string;
    score: number;
    start_a: number;
    end_a: number;
    start_b: number;
    end_b: number;
    operations: string[];
  };
};

const postJson = <TData>(url: string, body: unknown) => {
  return axiosInstance<TData>(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });
};

export const alignSequences = (payload: AlignmentAlignRequest) => {
  return postJson<AlignmentAlignResponse>("/alignment/align", payload);
};

export const useAlignmentAlignQuery = (
  payload: AlignmentAlignRequest | null,
  enabled = true,
): UseQueryResult<AlignmentAlignResponse, Error> => {
  return useQuery({
    queryKey: ["alignment", "align", payload],
    queryFn: () => {
      if (!payload) {
        throw new Error("Missing alignment request");
      }

      return alignSequences(payload);
    },
    enabled: enabled && Boolean(payload),
  });
};
